### ISSUE TITLE
REPL: Connexion refused in Archlinux

### ISSUE THREAD

**Issue by sshbio** [technomancy/leiningen#2030]
My config, in case this would be required:

```
$ uname -a
Linux arch 4.3.0-1-ARCH #1 SMP PREEMPT Mon Nov 2 16:52:35 CET 2015 x86_64 GNU/Linux
$ lein version
Leiningen 2.5.3 on Java 1.7.0_91 OpenJDK 64-Bit Server VM
$  java -version
java version "1.7.0_91"
OpenJDK Runtime Environment (IcedTea 2.6.2) (Arch Linux build 7.u91_2.6.2-1-x86_64)
OpenJDK 64-Bit Server VM (build 24.91-b01, mixed mode)
```

I had difficulties to connect to the REPL.  Here is how I get the issue:

```
$ lein new issue
$ cd issue
$ lein repl
Exception in thread "main" java.net.ConnectException: Connection refused, compiling:(/tmp/form-init7654738477480788732.clj:1:73)
	at clojure.lang.Compiler.load(Compiler.java:7239)
	at clojure.lang.Compiler.loadFile(Compiler.java:7165)
	at clojure.main$load_script.invoke(main.clj:275)
	at clojure.main$init_opt.invoke(main.clj:280)
	at clojure.main$initialize.invoke(main.clj:308)
	at clojure.main$null_opt.invoke(main.clj:343)
	at clojure.main$main.doInvoke(main.clj:421)
	at clojure.lang.RestFn.invoke(RestFn.java:421)
	at clojure.lang.Var.invoke(Var.java:383)
	at clojure.lang.AFn.applyToHelper(AFn.java:156)
	at clojure.lang.Var.applyTo(Var.java:700)
	at clojure.main.main(main.java:37)
Caused by: java.net.ConnectException: Connection refused
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:339)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:200)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:182)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:579)
	at java.net.Socket.connect(Socket.java:528)
	at java.net.Socket.<init>(Socket.java:425)
	at java.net.Socket.<init>(Socket.java:208)
	at clojure.tools.nrepl$connect.doInvoke(nrepl.clj:184)
	at clojure.lang.RestFn.invoke(RestFn.java:421)
	at clojure.tools.nrepl.ack$send_ack.invoke(ack.clj:47)
	at clojure.tools.nrepl.server$start_server.doInvoke(server.clj:157)
	at clojure.lang.RestFn.invoke(RestFn.java:619)
	at user$eval7514.invoke(form-init7654738477480788732.clj:1)
	at clojure.lang.Compiler.eval(Compiler.java:6782)
	at clojure.lang.Compiler.eval(Compiler.java:6772)
	at clojure.lang.Compiler.eval(Compiler.java:6772)
	at clojure.lang.Compiler.load(Compiler.java:7227)
	... 11 more
REPL server launch timed out.
```

As advised on the issue #1321, I can start lein headless and connect to it.

```
$ lein :headless
nREPL server started on port 36604 on host 127.0.0.1 - nrepl://127.0.0.1:36604
$ lein repl :connect nrepl://127.0.0.1:36604
Connecting to nREPL at nrepl://127.0.0.1:36604
REPL-y 0.3.7, nREPL 0.2.10
Clojure 1.7.0
OpenJDK 64-Bit Server VM 1.7.0_91-b01
    Docs: (doc function-name-here)
          (find-doc "part-of-name-here")
  Source: (source function-name-here)
 Javadoc: (javadoc java-object-or-class-here)
    Exit: Control+D or (exit) or (quit)
 Results: Stored in vars *1, *2, *3, an exception in *e

user=> 
```

I asked on #leiningen, and as advised, I ran jstack on java:
```
$ ps ax | grep java
 3556 pts/3    Sl+    0:30 java -Xbootclasspath/a:/home/josuah/.lein/self-installs/le
 4886 pts/4    Ss+    0:00 /usr/bin/grep java
$ jstack 3556
2015-11-18 04:14:39
Full thread dump OpenJDK 64-Bit Server VM (24.91-b01 mixed mode):

"Attach Listener" daemon prio=10 tid=0x00007fdae4001000 nid=0x11ca runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"ClosingPipeCloser 360581176" daemon prio=10 tid=0x00007fdb08393800 nid=0xe38 in Object.wait() [0x00007fdada5c0000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000f52ecb60> (a java.lang.Thread)
	at java.lang.Thread.join(Thread.java:1281)
	- locked <0x00000000f52ecb60> (a java.lang.Thread)
	at java.lang.Thread.join(Thread.java:1355)
	at com.hypirion.io.Pipe.join(Pipe.java:120)
	at com.hypirion.io.ClosingPipe$OutCloser.run(ClosingPipe.java:111)
	at java.lang.Thread.run(Thread.java:745)

"PipeThread 360581176" daemon prio=10 tid=0x00007fdb081b3000 nid=0xe37 in Object.wait() [0x00007fdada6c1000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000f52c9640> (a java.lang.Object)
	at java.lang.Object.wait(Object.java:503)
	at com.hypirion.io.RevivableInputStream.read(RevivableInputStream.java:140)
	- locked <0x00000000f52c9640> (a java.lang.Object)
	- locked <0x00000000f52c95b8> (a com.hypirion.io.RevivableInputStream)
	at java.io.InputStream.read(InputStream.java:101)
	at com.hypirion.io.Pipe$PipeOutputStream.run(Pipe.java:239)
	at java.lang.Thread.run(Thread.java:745)

"PipeThread 1815349000" daemon prio=10 tid=0x00007fdb0854f000 nid=0xe35 runnable [0x00007fdada7c2000]
   java.lang.Thread.State: RUNNABLE
	at java.io.FileInputStream.readBytes(Native Method)
	at java.io.FileInputStream.read(FileInputStream.java:272)
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:235)
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:275)
	at java.io.BufferedInputStream.read(BufferedInputStream.java:334)
	- locked <0x00000000f52e53a8> (a java.lang.UNIXProcess$ProcessPipeInputStream)
	at java.io.FilterInputStream.read(FilterInputStream.java:107)
	at com.hypirion.io.Pipe$PipeOutputStream.run(Pipe.java:239)
	at java.lang.Thread.run(Thread.java:745)

"PipeThread 1870807730" daemon prio=10 tid=0x00007fdb08576800 nid=0xe34 runnable [0x00007fdada8c3000]
   java.lang.Thread.State: RUNNABLE
	at java.io.FileInputStream.readBytes(Native Method)
	at java.io.FileInputStream.read(FileInputStream.java:272)
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:235)
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:275)
	at java.io.BufferedInputStream.read(BufferedInputStream.java:334)
	- locked <0x00000000f52e32e0> (a java.lang.UNIXProcess$ProcessPipeInputStream)
	at java.io.FilterInputStream.read(FilterInputStream.java:107)
	at com.hypirion.io.Pipe$PipeOutputStream.run(Pipe.java:239)
	at java.lang.Thread.run(Thread.java:745)

"process reaper" daemon prio=10 tid=0x00007fdb086ae800 nid=0xe33 runnable [0x00007fdb00061000]
   java.lang.Thread.State: RUNNABLE
	at java.lang.UNIXProcess.waitForProcessExit(Native Method)
	at java.lang.UNIXProcess.access$500(UNIXProcess.java:55)
	at java.lang.UNIXProcess$4.run(UNIXProcess.java:228)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

"RevivableReader 1254238582" daemon prio=10 tid=0x00007fdb0837c000 nid=0xe31 runnable [0x00007fdada9c4000]
   java.lang.Thread.State: RUNNABLE
	at java.io.FileInputStream.readBytes(Native Method)
	at java.io.FileInputStream.read(FileInputStream.java:243)
	at com.hypirion.io.RevivableInputStream$ThreadReader.run(RevivableInputStream.java:232)
	at java.lang.Thread.run(Thread.java:745)

"Service Thread" daemon prio=10 tid=0x00007fdb080b6800 nid=0xdee runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C1 CompilerThread1" daemon prio=10 tid=0x00007fdb080b4000 nid=0xded waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread0" daemon prio=10 tid=0x00007fdb080b1000 nid=0xdec waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Signal Dispatcher" daemon prio=10 tid=0x00007fdb080af000 nid=0xdeb runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Finalizer" daemon prio=10 tid=0x00007fdb08081800 nid=0xdea in Object.wait() [0x00007fdb01efd000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000cdab46f0> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:135)
	- locked <0x00000000cdab46f0> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:151)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:209)

"Reference Handler" daemon prio=10 tid=0x00007fdb0807f800 nid=0xde9 in Object.wait() [0x00007fdb01ffe000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000cdab4458> (a java.lang.ref.Reference$Lock)
	at java.lang.Object.wait(Object.java:503)
	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:133)
	- locked <0x00000000cdab4458> (a java.lang.ref.Reference$Lock)

"main" prio=10 tid=0x00007fdb08009000 nid=0xde5 in Object.wait() [0x00007fdb12469000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000f52ead70> (a java.lang.Thread)
	at java.lang.Thread.join(Thread.java:1281)
	- locked <0x00000000f52ead70> (a java.lang.Thread)
	at java.lang.Thread.join(Thread.java:1355)
	at com.hypirion.io.Pipe.join(Pipe.java:120)
	at leiningen.core.eval$sh.doInvoke(eval.clj:183)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invoke(core.clj:630)
	at leiningen.core.eval$fn__5830.invoke(eval.clj:253)
	at clojure.lang.MultiFn.invoke(MultiFn.java:233)
	at leiningen.core.eval$eval_in_project.invoke(eval.clj:356)
	at clojure.lang.AFn.applyToHelper(AFn.java:160)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invoke(core.clj:632)
	at leiningen.repl$repl.doInvoke(repl.clj:322)
	at clojure.lang.RestFn.invoke(RestFn.java:425)
	at clojure.lang.Var.invoke(Var.java:383)
	at clojure.lang.AFn.applyToHelper(AFn.java:156)
	at clojure.lang.Var.applyTo(Var.java:700)
	at clojure.core$apply.invoke(core.clj:632)
	at leiningen.core.main$partial_task$fn__6030.doInvoke(main.clj:261)
	at clojure.lang.RestFn.applyTo(RestFn.java:139)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:29)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invoke(core.clj:632)
	at leiningen.core.main$apply_task.invoke(main.clj:311)
	at leiningen.core.main$resolve_and_apply.invoke(main.clj:317)
	at leiningen.core.main$_main$fn__6096.invoke(main.clj:390)
	at leiningen.core.main$_main.doInvoke(main.clj:383)
	at clojure.lang.RestFn.invoke(RestFn.java:421)
	at clojure.lang.Var.invoke(Var.java:383)
	at clojure.lang.AFn.applyToHelper(AFn.java:156)
	at clojure.lang.Var.applyTo(Var.java:700)
	at clojure.core$apply.invoke(core.clj:630)
	at clojure.main$main_opt.invoke(main.clj:316)
	at clojure.main$main.doInvoke(main.clj:421)
	at clojure.lang.RestFn.invoke(RestFn.java:457)
	at clojure.lang.Var.invoke(Var.java:394)
	at clojure.lang.AFn.applyToHelper(AFn.java:165)
	at clojure.lang.Var.applyTo(Var.java:700)
	at clojure.main.main(main.java:37)

"VM Thread" prio=10 tid=0x00007fdb0807b800 nid=0xde8 runnable 

"GC task thread#0 (ParallelGC)" prio=10 tid=0x00007fdb0801e800 nid=0xde6 runnable 

"GC task thread#1 (ParallelGC)" prio=10 tid=0x00007fdb08020800 nid=0xde7 runnable 

"VM Periodic Task Thread" prio=10 tid=0x00007fdb080c1000 nid=0xdef waiting on condition 

JNI global references: 192
```

The person that gave me precious help from IRC suggested that the place it is stuck could be `com.hypirion.io.Pipe.join`.

Also, as suggested, I tried `lein run -m clojure.main`, which successfully started a REPL:

```
$ lein run -m clojure.main
Clojure 1.7.0
user=> 
```

Finally, he suggested that the facts that a "vanilla" repl works, but not an nrepl, means it could be OS security about socket usage.

Since technomancy is emacsen, I also put the content of `*Messages*` while attempting to connect with `M-x cider-jack-in`:

```
Making nrepl-create-client-buffer-function local to *nrepl-server issue* while let-bound!
Making nrepl-use-this-as-repl-buffer local to *nrepl-server issue* while let-bound!
Starting nREPL server via lein repl :headless...
nREPL server started on 33012
nREPL: Establishing direct connection to localhost:33012 ...
error in process filter: if: nREPL: Direct connection failed
error in process filter: nREPL: Direct connection failed
```

I remember that I have accidentally removed my linux user from all groups, while trying to add my user to audio group to configure jack2 for overtone, but I added it back to the wheel group.  I will reinstall Archlinux and try it again.

If needed, I can provide other informations and try whatever could fix it or provide required information.

Thank you for reading these (numerous) lines!

**Comment by hyPiRion** [comment]
It's unlikely that `com.hypirion.io.Pipe.join` would be the point where this is an issue – that code just redirects stdin/-out/-err to the subprocess. If that was an issue, then `lein run -m clojure.main` would've also failed.

It seems more likely that this is an issue related to network stuff. If this problem still persists after you'd reinstalled archlinux, I'd like you to try out `lein repl :connect nrepl://localhost:$port` instead of `lein repl :connect nrepl://127.0.0.1:$port` to verify that it's not some strange issue related to how localhost is resolved.

**Comment by sshbio** [comment]
Thank you, I tested with:

```
$ lein repl :headless
nREPL server started on port 33751 on host 127.0.0.1 - nrepl://127.0.0.1:33751
$ lein repl :connect nrepl://localhost:33751
```

And this provoked this exception:

```
Connecting to nREPL at nrepl://localhost:33751
ConnectException Connection refused
	java.net.PlainSocketImpl.socketConnect (PlainSocketImpl.java:-2)
	java.net.AbstractPlainSocketImpl.doConnect (AbstractPlainSocketImpl.java:339)
	java.net.AbstractPlainSocketImpl.connectToAddress (AbstractPlainSocketImpl.java:200)
	java.net.AbstractPlainSocketImpl.connect (AbstractPlainSocketImpl.java:182)
	java.net.SocksSocketImpl.connect (SocksSocketImpl.java:392)
	java.net.Socket.connect (Socket.java:579)
	java.net.Socket.connect (Socket.java:528)
	java.net.Socket.<init> (Socket.java:425)
	java.net.Socket.<init> (Socket.java:208)
	clojure.tools.nrepl/connect (nrepl.clj:184)
	clojure.core/apply (core.clj:630)
	clojure.tools.nrepl/add-socket-connect-method!/fn--10338 (nrepl.clj:224)
Bye for now!
```

With 127.0.0.1, it works:

```
$ lein repl :connect nrepl://127.0.0.1:33751
Connecting to nREPL at nrepl://127.0.0.1:33751
REPL-y 0.3.7, nREPL 0.2.10
Clojure 1.7.0
OpenJDK 64-Bit Server VM 1.7.0_91-b01
    Docs: (doc function-name-here)
          (find-doc "part-of-name-here")
  Source: (source function-name-here)
 Javadoc: (javadoc java-object-or-class-here)
    Exit: Control+D or (exit) or (quit)
 Results: Stored in vars *1, *2, *3, an exception in *e

user=> 
```

**Comment by hyPiRion** [comment]
I would guess you might be able to resolve this by adding the line `127.0.0.1	localhost` to the _top_ of `/etc/hosts` file. If the line's already in the file, just move it to the top.

I got the same bug when I removed the `127.0.0.1` entry and only had an IPv6 entry in there.

**Comment by hyPiRion** [comment]
Also while I'm at it, this seems to be related to #2000.

**Comment by sshbio** [comment]
Thank you for your proposition.

I had `127.0.0.1	localhost.localdomain	archbang arch` at the top, and I added `127.0.0.1 localhost` and rebooted, and it now works:

```
$ lein repl
nREPL server started on port 37017 on host 127.0.0.1 - nrepl://127.0.0.1:37017
REPL-y 0.3.7, nREPL 0.2.10
Clojure 1.7.0
OpenJDK 64-Bit Server VM 1.7.0_91-b01
    Docs: (doc function-name-here)
          (find-doc "part-of-name-here")
  Source: (source function-name-here)
 Javadoc: (javadoc java-object-or-class-here)
    Exit: Control+D or (exit) or (quit)
 Results: Stored in vars *1, *2, *3, an exception in *e

user=> 
```

It also worked with the emacs cider plugin.

I do not know if this is archlinux-or leiningen-related issue.

You found what the issue was at first try!

Maybe you can close the issue.  Thank you

**Comment by hyPiRion** [comment]
I guess the solution to #2000 would also resolve the issue for this case "for free". I'll merge this into that one.
